### PR TITLE
Working towards release 1.0

### LINKED ISSUE #96: Downsampling of cells after filtering
Das Donwsampling würden wir gerne nicht nur zu Testzwecken, sondern als festen Bestandteil in das Skript integrieren, um die Anzahl der Zellen pro Sample an einander anpassen zu können und im Folgenden mit den identischen Zellzahlen weiterarbeiten zu können. Hierfür wäre es dann sinnvoller, das Downsampling hinter das Filtern zu verschieben.

### LINKED ISSUE #102: Small changes before release
I figured that for my projects it is helpful to plot QC again after we filtered based on the given thresholds. I'd like to include this into the main `scrnaseq.Rmd`. 

Edit 23.2.22: done

### PR DESCRIPTION
**Pull request by ktrns**
- Fixed #102 
- Fixed #96 
- Added downsampling option AFTER filtering so all samples have the same size
- Fixed figure sizes in case there are 20-30 clusters

### CODE REVIEW

File: scrnaseq.Rmd [hunk 1]
```diff
@@ -17,9 +17,14 @@ params:
...
-  # Downsample data to at most n cells per sample (mainly for tests)
+  # Downsample data to at most n cells per sample AFTER filtering (mainly for tests)
   #   NULL to deactivate
-  downsample_cells_n: 100
+  downsample_cells_n: NULL
+  
+  # Downsample all samples equally according to the smalles sample
+  #   TRUE/FALSE
+  #   Overwritten by downsample_cells_n
+  downsample_cells_equally: TRUE
```
Comment (andpet0101) [inline scrnaseq.Rmd:27]: Should be set to FALSE.

File: scrnaseq.Rmd [hunk 2]
```diff
@@ -17,9 +17,14 @@ params:
   # Data type ("RNA", "Spatial")
   assay_raw: "RNA"
   
-  # Downsample data to at most n cells per sample (mainly for tests)
+  # Downsample data to at most n cells per sample AFTER filtering (mainly for tests)
   #   NULL to deactivate
-  downsample_cells_n: 100
+  downsample_cells_n: NULL
+  
+  # Downsample all samples equally according to the smalles sample
```
Comment (andpet0101) [inline scrnaseq.Rmd]: Typo

File: scrnaseq.Rmd [hunk 3]
```diff
@@ -1030,9 +1032,31 @@ if (length(p_list)==1) {
 }
 p
 ```
+```{r downsampling}
+# If requested, determine number of cells for downsampling
+if (is.null(param$downsample_cells_n) & param$downsample_cells_equally) {
+  n = purrr::map_int(sc, ncol) %>% min()  
+} else if (!is.null(param$downsample_cells_n)) {
+  n = param$downsample_cells_n
```
Comment (andpet0101) [inline scrnaseq.Rmd:1040]: What will happen if a sc object has less than n cells?

File: scrnaseq.Rmd [hunk 4]
```diff
@@ -2239,25 +2283,26 @@ if (markers_found) {
...
 
-  # Feature plots and violin plots: each row contains 5 plots
-  #   The plot has 5 columns and 1 row per cluster, hence the layout works nicely if we find 
-  #     at least 5 markers per cluster
-  nr_rows_5cols = ceiling(nrow(markers_top)/5)
-  fig_height_5cols = height_per_row * nr_rows_5cols
+  # Feature plots and violin plots: each row contains 3 plots
+  #   The plot has 3 columns and 2 rows per cluster, hence the layout works nicely if we find 
+  #     at least 6 markers per cluster
+  nr_rows_3cols = ceiling(nrow(markers_top)/3)
```
Comment (andpet0101) [inline scrnaseq.Rmd:2291]: I want a function ....

**Comment by andpet0101** [comment]
Okay a few little things ... please

**Review by andpet0101 — CHANGES_REQUESTED** [review]

File: scrnaseq.Rmd (hunk 1 above)
Comment (ktrns) [reply scrnaseq.Rmd:27]: Not yet. We had `downsample_cells_n` set to 100 for this test report as long as we don't have a final release. It's faster to compile for testing :-). So I just switched to this and set `downsample_cells_n=NULL`.

File: scrnaseq.Rmd (hunk 2 above)
Comment (ktrns) [reply scrnaseq.Rmd]: Done

File: scrnaseq.Rmd (hunk 3 above)
Comment (ktrns) [reply scrnaseq.Rmd:1040]: You take care of that in `ScSampleCells`: 
```
  # Set n
  if (n > ncol(sc)) n = ncol(sc)
  ```

File: scrnaseq.Rmd (hunk 4 above)
Comment (ktrns) [reply scrnaseq.Rmd:2291]: Go for it, not in my PR though :-)

File: scrnaseq.Rmd (hunk 1 above)
Comment (andpet0101) [reply scrnaseq.Rmd:27]: Okay so you rely on the test dataset having 100 cells ?

File: scrnaseq.Rmd (hunk 3 above)
Comment (andpet0101) [reply scrnaseq.Rmd:1040]: yes true

File: scrnaseq.Rmd (hunk 4 above)
Comment (andpet0101) [reply scrnaseq.Rmd:2291]: I will add it to the project board :-)

File: scrnaseq.Rmd (hunk 1 above)
Comment (andpet0101) [reply scrnaseq.Rmd:27]: okay

**Review by andpet0101 — APPROVED** [review]
done

**Pull request merged**